lib/Dialect/P4HIR/P4HIR_Ops.cpp

```diff
@@ -2281,12 +2281,168 @@ void P4HIR::SwitchOp::build(OpBuilder &builder, OperationState &result, mlir::Va
                             function_ref<void(OpBuilder &, Location)> switchBuilder) {
     assert(switchBuilder && "the builder callback for regions must be present");
     OpBuilder::InsertionGuard guardSwitch(builder);
-    Region *swtichRegion = result.addRegion();
-    builder.createBlock(swtichRegion);
+    Region *switchRegion = result.addRegion();
+    builder.createBlock(switchRegion);
     result.addOperands(cond);
     switchBuilder(builder, result.location);
 }
 
+//===----------------------------------------------------------------------===//
+// ForOp
+//===----------------------------------------------------------------------===//
+
+void P4HIR::ForOp::build(
+    OpBuilder &builder, OperationState &result, mlir::DictionaryAttr annotations,
+    llvm::function_ref<void(mlir::OpBuilder &, mlir::Location)> condBuilder,
+    llvm::function_ref<void(mlir::OpBuilder &, mlir::Location)> bodyBuilder,
+    llvm::function_ref<void(mlir::OpBuilder &, mlir::Location)> updateBuilder) {
+    OpBuilder::InsertionGuard guard(builder);
+
+    Region *condRegion = result.addRegion();
+    builder.createBlock(condRegion);
+    condBuilder(builder, result.location);
+
+    if (annotations && !annotations.empty())
+        result.addAttribute(getAnnotationsAttrName(result.name), annotations);
+
+    Region *bodyRegion = result.addRegion();
+    builder.createBlock(bodyRegion);
+    bodyBuilder(builder, result.location);
+
+    Region *updateRegion = result.addRegion();
+    builder.createBlock(updateRegion);
+    updateBuilder(builder, result.location);
+}
+
+void P4HIR::ForOp::print(mlir::OpAsmPrinter &p) {
```

Comment:
what's the problem with ODS-generated printer? Why can't you use e.g. 
```tablegen
(`annotations` $annotations^)?
``` 
there?

Reply:
Right. Unparsed annotations are just there as-is. You'd want to use e.g. `@likely` :)